components/launcher/src/server/mod.rs

```diff
@@ -208,7 +208,32 @@ impl Server {
                 );
             }
         }
-        self.supervisor.wait().ok();
+
+        // With the Supervisor shutting down services, we need to
+        // ensure that the Launcher is able to reaping those main
+        // service processes as the Supervisor shuts them
+        // down... otherwise, the Supervisor can end up waiting a long
+        // time on zombie processes.
+        loop {
+            self.services.reap_services();
+            match self.supervisor.try_wait() {
+                Ok(Some(status)) => {
+                    debug!("Supervisor exited with: {}", status);
+                    break;
+                }
+                Ok(None) => {
+                    // Still hasn't exited; try again later
+                }
+                Err(e) => {
+                    error!("Error waiting on supervisor: {:?}", e);
+                    break;
+                }
+            }
+        }
```

Comment:
From my reading of this and [reap_services](https://github.com/habitat-sh/habitat/blob/master/components/launcher/src/server/mod.rs#L382-L405), I don't see any correctness issues.  Provided that's true, I prefer @baumanj's version.

Reply:
We _do_ need continuous reaping right now, but I'm digging further into our `is_alive` function to see if that needs to be fixed instead. If that's the case, we might be able to go with @baumanj's suggestion.

Reply:
http://man7.org/linux/man-pages/man2/kill.2.html

> ESRCH  The process or process group does not exist.  Note that an
              existing process might be a zombie, a process that has
              terminated execution, but has not yet been wait(2)ed for.

Currently our `is_alive` function is just a fancy call to `libc::kill`, and thus has no way to distinguish between a process that is alive and one that is a zombie. As a result, until the Launcher reaps it, the Supervisor is going to wait its 8 seconds before sending it a `KILL`, which will needlessly eat up time in the Supervisor. As we move to longer customizable timeouts, this problem stands to only get worse.

Slightly complicating things is the fact that we're often using the _negative_ PID, which is actually trying to see if any process in the process _group_ is alive (we should probably call it `exists` rather than `is_alive`).

We could peek into `/proc` with [procinfo](https://docs.rs/procinfo/0.4.2/procinfo/pid/enum.State.html) to determine if the process is alive or a zombie, but we'd need to modify things to look at all the members of the process group (since `procinfo` isn't going to do much with negative pids).

All these are probably worthwhile changes to make, but seem to be their own separate thing. In the mean time, waiting on the services as we shut down can achieve the same effect.

Thoughts?

Reply:
Given all that, my opinion would be to leave it for now, since it does work, and file an issue to make those improvements.

Reply:
I'm still gonna steal @baumanj's suggestion, since it's pretty slick. 

Issue filed for refactoring `is_alive`: #6131